app/src/main/java/com/feedhenry/securenativeandroidtemplate/AuthenticationFragment.java

```diff
@@ -5,200 +5,58 @@
-            // perform the user login attempt.
-            showProgress(true);
-            mAuthTask = new UserLoginTask(username, password);
-            mAuthTask.execute((Void) null);
-        }
-    }
-
-    /**
-     * Shows the progress UI and hides the login form.
-     */
-    private void showProgress(final boolean show) {
-        mProgressView.setVisibility(show ? View.VISIBLE : View.GONE);
-        mLoginFormView.setVisibility(show ? View.GONE : View.VISIBLE);
-    }
-
-    /**
-     * Allow override the default auth provider. Mainly used for tests.
-     * @param authProvier
-     */
-    public void setAuthProvider(AuthenticateProvider authProvier) {
-        mAuthProvider = authProvier;
-    }
-
-
-    /**
-     * Set a callback that will be invoked when the authentication is completely successfully
-     * @param callback
-     */
-    public void setAuthSuccessCallback(AuthenticateSuccessCallback callback) {
-        mAuthSuccessCallback = callback;
-    }
-
-    /**
-     * Represents an asynchronous login/registration task used to authenticate
-     * the user.
+     * Login via Keycloak with a redirect to the Native browser on the device.
      */
-    public class UserLoginTask extends AsyncTask<Void, Void, AuthenticateResult> {
-
-        private final String mUsername;
-        private final String mPassword;
-
-        UserLoginTask(String username, String password) {
-            mUsername = username;
-            mPassword = password;
-        }
-
-        /**
-         * Placeholder for performing authentication in the background
-         */
-        @Override
-        protected AuthenticateResult doInBackground(Void... params) {
-            return mAuthProvider.authenticateWithUsernameAndPassword(mUsername, mPassword);
-        }
-
-        /**
-         * Method to handle the UI after authentication has completed
-         */
-        @Override
-        protected void onPostExecute(final AuthenticateResult result) {
-            mAuthTask = null;
-            showProgress(false);
-
-            if (result != null) {
-                if (mAuthSuccessCallback != null) {
-                    mAuthSuccessCallback.authenticated(result);
-                }
-            } else {
-                mPasswordView.setError(getString(R.string.error_incorrect_credentials));
-                mPasswordView.requestFocus();
-            }
-        }
+    public void attemptKeycloakLogin() {
+        ((MainActivity)getActivity()).performKeycloakAuthentication();
     }
 
     /**
-     * To allow the caller define what actions need to be performed when the authentication is completed
+     * Login via Keycloak with a redirect to the Native browser on the device.
      */
-    public interface AuthenticateSuccessCallback {
-        public void authenticated(AuthenticateResult result);
+    public void attemptKeycloakLogout() {
+        ((MainActivity)getActivity()).performKeycloakLogout();
```

Comment:
invoke the functions in the `MainActivity` from the fragment will work, but the biggest problem here is introducing the dependency on the `MainActivity`, thus make it hard to reuse the fragment somewhere else.

A better way to do it is to use callbacks interfaces, and let the caller to implement the interfaces and decide what then want to do.

For example, in here we can declare an interface: `OnAuthenticationActions`, it will have 2 methods: `performLogin` and `performLogout`.  This class will then invoke the functions of the interface when the buttons are clicked.

Then from the `MainActivity`,  it will create a new implementation of the interface and set it on the fragment. And the caller can decide what it want to do when the buttons are clicked.

Does this make sense?

Reply:
or check the second answer in SO: https://stackoverflow.com/questions/12659747/call-an-activity-method-from-a-fragment